gdal/alg/gdalrasterize.cpp

```diff
@@ -78,56 +89,128 @@ void gvBurnScanline( void *pCBData, int nY, int nXStart, int nXEnd,
     if( nXEnd >= psInfo->nXSize )
         nXEnd = psInfo->nXSize - 1;
 
-    if( psInfo->eType == GDT_Byte )
+    for( int iBand = 0; iBand < psInfo->nBands; iBand++ )
     {
-        for( int iBand = 0; iBand < psInfo->nBands; iBand++ )
-        {
-            unsigned char nBurnValue = static_cast<unsigned char>
-                ( psInfo->padfBurnValue[iBand] +
-                  ( (psInfo->eBurnValueSource == GBV_UserBurnValue)?
-                             0 : dfVariant ) );
-
-            unsigned char *pabyInsert =
-                psInfo->pabyChunkBuf
-                + iBand * psInfo->nXSize * psInfo->nYSize
-                + nY * psInfo->nXSize + nXStart;
-
-            if( psInfo->eMergeAlg == GRMA_Add ) {
-                int nPixels = nXEnd - nXStart + 1;
-                while( nPixels-- > 0 )
-                    *(pabyInsert++) += nBurnValue;
-            } else {
-                memset( pabyInsert, nBurnValue, nXEnd - nXStart + 1 );
+        const double burnValue = ( psInfo->padfBurnValue[iBand] +
+                ( (psInfo->eBurnValueSource == GBV_UserBurnValue)?
+                            0 : dfVariant ) );
+
+        unsigned char *pabyInsert = psInfo->pabyChunkBuf
+                                    + iBand * psInfo->nBandSpace
+                                    + nY * psInfo->nLineSpace + nXStart * psInfo->nPixelSpace;
+        int nPixels = nXEnd - nXStart + 1;
+        if( psInfo->eMergeAlg == GRMA_Add ) {
+            switch (psInfo->eType)
+            {
+                case GDT_Byte:
```

Comment:
Unless I'm missing something this seems begging for a template, as does the code below.  A template would make for less code and reduce likelihood of an error.